src/main/kotlin/org/rust/ide/typing/paste/RsImportCopyPasteProcessor.kt

```diff
@@ -19,25 +19,82 @@ import com.intellij.psi.PsiDocumentManager
 import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiFile
 import com.intellij.psi.impl.source.tree.injected.changesHandler.range
+import org.rust.ide.inspections.fixes.QualifyPathFix
 import org.rust.ide.inspections.import.AutoImportFix
 import org.rust.ide.settings.RsCodeInsightSettings
-import org.rust.ide.utils.import.ImportCandidate
-import org.rust.ide.utils.import.import
+import org.rust.ide.utils.import.*
 import org.rust.lang.core.psi.*
-import org.rust.lang.core.psi.ext.RsElement
-import org.rust.lang.core.psi.ext.RsQualifiedNamedElement
-import org.rust.lang.core.psi.ext.qualifiedName
-import org.rust.lang.core.psi.ext.startOffset
+import org.rust.lang.core.psi.ext.*
 import org.rust.lang.core.types.inference
 import org.rust.openapiext.toPsiFile
 import java.awt.datatransfer.DataFlavor
 import java.awt.datatransfer.Transferable
 
-data class ImportMap(private val offsetToFqnMap: Map<Int, String>) {
-    fun elementToFqn(element: PsiElement, range: TextRange): String? {
-        val offset = toRelativeOffset(element, range)
-        return offsetToFqnMap[offset]
+data class QualifiedItemPath(val qualifiedName: String, val crateRelative: String)
+
+data class ElementRange(val start: Int, val length: Int)
```

Comment:
I am not quite understand, why can't we use single start offset as before?

Reply:
I stored the mapping keyed by the start offset. But there can be multiple elements at the same start offset, and they just overwrote each other. Before it worked by accident mostly, but after he changes from this PR it started causing problems in tests.

Reply:
What if we store just the end offset? Will it also cause collisions?

Reply:
I recall that I have tried it and it also caused collisions, I'll try again.

Reply:
All tests seem to pass if we just remember the end offset.